src/NLog/Targets/TargetWithContext.cs

```diff
@@ -0,0 +1,802 @@
+//   this list of conditions and the following disclaimer. 
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Targets
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Text;
+    using NLog.Config;
+    using NLog.Internal;
+    using NLog.Layouts;
+
+    /// <summary>
+    /// Represents target that supports context capture using MDLC, MDC, NDLC and NDC
+    /// </summary>
+    public abstract class TargetWithContext : TargetWithLayout, IIncludeContext
+    {
+        /// <summary>Internal Layout that allows capture of MDC context</summary>
+        public Layout InternalLayoutMdc { get; }
+        /// <summary>Internal Layout that allows capture of NDC context</summary>
+        public Layout InternalLayoutNdc { get; }
+#if !SILVERLIGHT
+        /// <summary>Internal Layout that allows capture of MDLC context</summary>
+        public Layout InternalLayoutMdlc { get; }
+        /// <summary>Internal Layout that allows capture of NDLC context</summary>
+        public Layout InternalLayoutNdlc { get; }
+#endif
+        /// <summary>Internal Layout that allows capture of callsite context</summary>
+        public Layout InternalLayoutCallSite { get; }
+
+        bool IIncludeContext.IncludeAllProperties { get => IncludeEventProperties; set => IncludeEventProperties = value; }
+
+        /// <inheritdoc/>
+        /// <docgen category='General Options' order='10' />
+        bool IncludeEventProperties
+        {
+            get { return _includeEventProperties && !_contextLayout; }
+            set
+            {
+                _includeEventProperties = value;
+                if (Layout is IIncludeContext contextLayout)
```

Comment:
I think it would be nice to always set the contextLayout, but it should be a fake, then create a "fake" impl.

So all the properties just do   `contextLayout.IncludeAllProperties = value;`